aten/src/ATen/native/nested/NestedTensorMath.cpp

```diff
@@ -661,5 +661,50 @@ Tensor& dropout_nested_(Tensor& input, double p, bool train) {
   return input;
 }
 
+Tensor softmax_nested(const Tensor& input, const int64_t dim, const bool half_to_float) {
+  auto input_ptr = get_nested_tensor_impl(input);
+  int64_t positive_dim = at::maybe_wrap_dim(dim, input_ptr->dim());
+  TORCH_CHECK(
+      positive_dim >= 1,
+      "Cannot apply softmax across nested dimension 0");
+  const Tensor & buffer = input_ptr->get_buffer(),
+                 sizemat = input_ptr->get_nested_size_tensor();
+  Tensor output_buffer = buffer.new_empty(buffer.sizes());
+  // split buffer into original tensors
+  int64_t ntensors = *(input_ptr->opt_size(0));
+  int64_t original_dim = sizemat.size(1);
+  std::vector<int64_t> starts(ntensors + 1);
+  starts[0] = 0;
+  starts.back() = buffer.numel();
+  std::vector<IntArrayRef> shapes(ntensors);
+  const int64_t * sizemat_ptr = sizemat.data_ptr<int64_t>();
+  int64_t sizemat_offset = 0;
+  for (int64_t i = 0; i < ntensors - 1; i++) {
+    shapes[i] = IntArrayRef(sizemat_ptr + sizemat_offset, sizemat_ptr + sizemat_offset + original_dim);
+    int64_t row_product = sizemat_ptr[sizemat_offset];
+    sizemat_offset++;
+    for (int64_t j = 1; j < original_dim; j++) {
+      row_product *= sizemat_ptr[sizemat_offset];
+      sizemat_offset++;
+    }
+    starts[i + 1] = starts[i] + row_product;
+  }
+  shapes.back() = IntArrayRef(sizemat_ptr + sizemat_offset, sizemat_ptr + sizemat_offset + original_dim);
+  // call tensor softmax
+  // TODO: for cpu, maybe use `parallel_for`
+  //       to do that, have to merge `aten/src/ATen/native/cpu/SoftMaxKernel.cpp/softmax_kernel`
+  //       1. it has `parallel_for` and we cannot multi-thread in multi-thread
+  //       2. cannot dispatch in multi-thread (in this case at::_softmax_out)
+  for (int64_t i = 0; i < ntensors; i++) {
```

Comment:
Since we're looping over the Tensor constituents we could also call into unbind right away. It might roughly look like the following.

```
std::vector<Tensor> result_tensors;
for (const auto& tensor : input.unbind()) {
    result_tensors.push_back(tensor.softmax(positive_dim - 1, half_to_float));
}
return at::nested_tensor(result_tensors);
```

Reply:
This looks clean, but I think it does slightly more copying than the implementation in this PR, which directly writes to a pre-allocated output tensor. There's also the added dispatcher overhead of calling `unbind()` rather than a util to get the constituents. I'd expect the former to have a bigger impact to perf in practice, and maybe the latter doesn't matter